### PR TITLE
[Perf] Qwen3-Omni performance optimization

### PR DESCRIPTION
**Pull request by amy-why-3459**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

1. Talker and Code2WAV skip `get_mrope_input_positions`
image
2. Add `("codes","audio")` to `gpu_resident_buffer_keys` in `qwen3_omni.py`
3. Change `_store_value` to pinned memory + non-blocking D2H
ba71ec88b11946e708ad6e40ed19457b
4. Asynchronously prefetch `sample_tokens` hidden states to a secondary CUDA stream
06aac5bbe19fe30afa7d044113ce436f

## Test Plan

`pytest -s -v tests/dfx/perf/scripts/run_benchmark.py --test-config-file tests/dfx/perf/tests/test_qwen3_omni_async_chunk.json`

## Test Result

concurrency | RTF(main) | RTF(PR 3878) | delta
-- | -- | -- | --
1 | 0.14 | 0.13 | -7.14%
4 | 0.25 | 0.22 | -12.00%
8 | 0.31 | 0.26 | -16.13%
16 | 0.43 | 0.35 | -18.60%
32 | 0.67 | 0.50 | -25.37%

concurrency | req/s(main) | req/s(PR 3878) | delta
-- | -- | -- | --
1 | 0.05 | 0.04 | -20.00%
4 | 0.1 | 0.11 | 10.00%
8 | 0.17 | 0.17 | 0.00%
16 | 0.24 | 0.27 | 12.50%
32 | 0.29 | 0.35 | 20.69%

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
--- Running Summary
=========== 36 passed, 3 skipped, 17 warnings in 2897.66s (0:48:17) ============
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
any acc comparison?

**Comment by amy-why-3459** [comment]
> any acc comparison?

Please wait, the test cases are running.

**Comment by amy-why-3459** [comment]
> any acc comparison?

Updated

```
pytest -sv tests/e2e/online_serving/test_qwen3_omni_expansion.py -m "full_model" --run-level "full_model"

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
--- Running Summary
=========== 36 passed, 3 skipped, 17 warnings in 2897.66s (0:48:17) ============
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```

```
===== Seed-TTS eval (seed-tts-eval protocol) =====
Evaluated (WER, lower is better):        1088      
Mean WER:                                0.4321    
Median WER:                              0.0000    
Request failed:                          0         
No PCM captured:                         0         
ASR / WER failed:                        0         
==================================================
```

**Comment by gcanlin** [comment]
Could you also show the comparison of profiling between before and after?

**Comment by amy-why-3459** [comment]
> Could you also show the comparison of profiling between before and after?

okay
1. Stage-2: 159ms -> 587us
image
2. Stage-1: 5ms -> 19us
image
3. The third optimization is not very obvious; further optimization will be carried out.
Thirdly, I think a more long-term modification plan needs to be optimized from the connector side.

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py [hunk 1]
```diff
@@ -167,6 +166,20 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
                 hf_config=talker_config,
                 architectures=["Qwen3OmniMoeTalkerForConditionalGeneration"],
             )
+            # Be tolerant to talker implementations loaded from different
+            # package snapshots (repo vs installed wheel) that may expose
+            # either init_multi_modal or init_multimodal, or neither.
+            init_mm = getattr(self.talker, "init_multi_modal", None)
```
Comment (ZeldaHuang) [inline vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py]: Qwen3 Omni talker models don't need multimodal initialization — it seems a dead code that only triggers a warning.

**Review by hsliuustc0106 — COMMENTED** [review]
The benchmark data is strong (5 concurrency levels, clear RTF/reqs tables, runnable script, pytest summary). A few things to add to the PR body:

- Hardware spec: GPU model, count, VRAM ("tested on GPU" is not enough)
- Software versions: torch, CUDA, vllm, vllm-omni versions
- Warmup and measurement count: "1 warmup + 3 measured" or equivalent, plus whether values are averages or single runs
- VRAM: peak GPU memory before and after
- Concurrency=1 regression: req/s drops 20 percent at concurrency=1 — expected noise, tradeoff, or something else?

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py (hunk 1 above)
Comment (amy-why-3459) [reply vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py]: > Qwen3 Omni talker models don't need multimodal initialization — it seems a dead code that only triggers a warning.

Yes, fixed

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py [hunk 2]
```diff
@@ -237,22 +239,14 @@ async def buffer_realtime_audio(
 
         prompt_token_ids = tokenizer.encode(prompt_template)
 
-        # In non-async-chunk (full-payload) mode the engine treats each
-        # streaming TokensPrompt as a fresh decode, so mid-stream segment
-        # yields cause the thinker to emit duplicate responses and the
-        # talker to emit duplicate audio. Defer all audio to the final
-        # flush so the thinker sees one complete prompt.
-        async_chunk = getattr(model_config, "async_chunk", False)
```
Comment (ZeldaHuang) [inline vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py:245]: Will this deletion affect non-async scenario?

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py [hunk 3]
```diff
@@ -323,7 +317,15 @@ def get_mrope_input_positions(
                 msg = "Qwen3 Omni thinker get_mrope_input_positions requires mm_features"
                 raise ValueError(msg)
             return self.thinker.get_mrope_input_positions(input_tokens, mm_features)
-        return MRotaryEmbedding.get_input_positions_tensor(input_tokens, **kwargs)
+        # Talker/code2wav stages are text/codec-only and do not need
+        # multimodal M-RoPE position computation. Return a cheap linear
+        # position tensor to avoid unnecessary per-request M-RoPE work.
+        seq_len = len(input_tokens)
```
Comment (ZeldaHuang) [inline vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py:331]: Maybe a simpler version:`linear = torch.arange(seq_len, dtype=torch.long).unsqueeze(0).expand(3, seq_len)
`

File: vllm_omni/worker/gpu_ar_model_runner.py [hunk 4]
```diff
@@ -95,6 +95,19 @@ def __init__(self, *args, **kwargs):
                 kv_transfer_manager=self.kv_transfer_manager,
             )
         self._downstream_payload_cache: dict[str, bool] = {}
+        # Persistent D2H stream (created once, reused every step).
+        self._hs_d2h_stream: torch.cuda.Stream = torch.cuda.Stream()
```
Comment (ZeldaHuang) [inline vllm_omni/worker/gpu_ar_model_runner.py]: Is it worth using a separate stream for this D2H copy if it takes a few μs?(decode hidden_states usually <1MB and stream synchronization also has some overhead)

Comment (amy-why-3459) [reply vllm_omni/worker/gpu_ar_model_runner.py]: Based on profiling, D2H takes about 10ms. However, I think long-term optimization from the connector side is necessary.

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py (hunk 3 above)
Comment (amy-why-3459) [reply vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py:331]: > Maybe a simpler version:`linear = torch.arange(seq_len, dtype=torch.long).unsqueeze(0).expand(3, seq_len) `

Done

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py (hunk 2 above)
Comment (amy-why-3459) [reply vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py:245]: > Will this deletion affect non-async scenario?

Fixed

File: vllm_omni/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -1714,7 +1714,15 @@ def _store_value(self, dest: dict, key: str, value: Any, gpu_keys: set) -> None:
...
             else:
-                dest[key] = value.detach().to("cpu").contiguous()
+                t = value.detach()
+                if t.is_cuda:
+                    # Blocking D2H: after Plan 1 keeps codes.audio on GPU the
+                    # remaining tensors are small (< 100 KB), so blocking is
+                    # negligible (< 0.1 ms) and avoids any race between the
+                    # async DMA and the next step's _preprocess reading the buf.
+                    dest[key] = t.to("cpu").contiguous()
+                else:
```
Comment (ZhengWG) [inline vllm_omni/worker/gpu_model_runner.py:1771]: Just curious: is it for npu_branch?

Comment (amy-why-3459) [reply vllm_omni/worker/gpu_model_runner.py:1771]: > Just curious: is it for npu_branch?

Used to distinguish between CPU tensors and GPU tensors

**Comment by linyueqian** [comment]
Spot-checked this on Qwen3-TTS (CustomVoice, seed-tts-text, c=64 N=512 W=8 warmups, H20) since it touches the shared `gpu_ar_model_runner.py` / `gpu_model_runner.py` paths.

| | PR `68ced906` cold | main `abc31731` cold | Δ |
|---|---|---|---|
| median TTFP | 6003.6 ms | 6032.8 ms | -0.5% |
| median RTF | 1.259 | 1.275 | -1.3% |
| throughput | 50.0 | 49.6 | +0.8% |

Within ±3% bench noise, so no measurable effect on Qwen3-TTS. Consistent with expectation: techniques #1 (mrope skip) and #2 (`codes.audio` GPU-resident) are Qwen3-Omni-specific, and Qwen3-TTS already lists `("codes","audio")` in `gpu_resident_buffer_keys`. The shared-runner pieces (#3 pinned `_store_value`, #4 async hidden-state prefetch) apply transitively but evidently do not find an overlap window in the Qwen3-TTS hot path.

No regression for Qwen3-TTS, so safe to merge from a TTS perspective. Nice work on the Qwen3-Omni gains.

File: vllm_omni/worker/gpu_model_runner.py (hunk 5 above)
Comment (ZeldaHuang) [reply vllm_omni/worker/gpu_model_runner.py:1771]: > > Just curious: is it for npu_branch?
> 
> Used to distinguish between CPU tensors and GPU tensors

Will this affect performance? Previous implementation looks more concise.

Comment (amy-why-3459) [reply vllm_omni/worker/gpu_model_runner.py:1771]: > > > Just curious: is it for npu_branch?
> > 
> > 
> > Used to distinguish between CPU tensors and GPU tensors
> 
> Will this affect performance? Previous implementation looks more concise.

It affects performance; the time taken before the modification was 5ms, and the time taken after the modification was 19us.

Comment (ZeldaHuang) [reply vllm_omni/worker/gpu_model_runner.py:1771]: I'm a little confused about this performance gain—where does it come from?

Comment (amy-why-3459) [reply vllm_omni/worker/gpu_model_runner.py:1771]: > I'm a little confused about this performance gain—where does it come from?

**tensor.contiguous() (already on CPU)**
    If contiguous: returns self, O(1), no copy.
    If not contiguous: allocates a new buffer and copies; cost ≈ memory bandwidth for numel × element_size.
**tensor.to("cpu").contiguous() (already on CPU)**
    .to("cpu"): sees tensor is already on CPU → usually returns the same tensor (no D2H).
    .contiguous(): same as above.
So the extra cost is almost entirely **.to() dispatch overhead** (device/dtype/layout checks), not a second copy.

image

Comment (ZeldaHuang) [reply vllm_omni/worker/gpu_model_runner.py:1771]: So the gain should be tiny (~us), 5ms->19us may comes from other part?

**Comment by amy-why-3459** [comment]
Since pinned memory still requires synchronization, I have rolled back the changes. A new solution will be implemented after @natureofnature 's refactor communication layer is completed.

H100:
concurrency | RTF(main) | RTF(PR 3878) | delta
-- | -- | -- | --
1 | 0.14 | 0.13 | -7.14%
4 | 0.25 | 0.23 | -8.00%
8 | 0.32 | 0.29 | -9.38%
16 | 0.45 | 0.39 | -13.33%
32 | 0.69 | 0.60 | -13.04%

**Review by ZeldaHuang — APPROVED** [review]

File: vllm_omni/worker/gpu_model_runner.py (hunk 5 above)
Comment (ZhengWG) [reply vllm_omni/worker/gpu_model_runner.py:1771]: > So the gain should be tiny (~us), 5ms->19us may comes from other part?

Agreed, it's a little confusing here.

Comment (amy-why-3459) [reply vllm_omni/worker/gpu_model_runner.py:1771]: > > So the gain should be tiny (~us), 5ms->19us may comes from other part?
> 
> Agreed, it's a little confusing here.

I updated the comments; the average execution time of `_store_value` decreased from 249µs to 28µs. The improvement in `_store_value`'s time efficiency comes primarily from two aspects:
1. Talker MTP codec codes must remain on the GPU to avoid per-step D2H.
2. If the tensor is already on the CPU, it doesn't need to be offloaded to the CPU, thus reducing CPU scheduling.
24bc2656004ffc9a4f302db90ec85c25

**Comment by amy-why-3459** [comment]
CI's failure is unrelated to this PR.
